File: leetcode/001_two_sum/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    var ans = [];
    var myMap = {};
    nums.forEach(function(num, index){
        myMap[num] ? myMap[num].push(index) : myMap[num] = [index];
    });
    var isBreak = false;
    Object.keys(myMap).forEach(function(key){
        if(isBreak)return;
        var temp = target - key;
        if(myMap[temp]){
            ans.push(myMap[key][0]);
            ans.push(temp == key ? myMap[key][1] : myMap[temp][0]);
            isBreak = true;
        }
    });
    ans.sort();
    return ans;
};

var nums = [2, 7, 11, 15];
var target = 9;
// var nums = [3,3];
// var target = 6;
console.log(twoSum(nums, target));

File: leetcode/001_two_sum/twoSum.go
// You can edit this code!
// Click here and start typing.
package main

import (
	"sort"
)

func twoSum(nums []int, target int) []int {
	result := make([]int, 2)
	myMap := make(map[int][]int)
	for idx, num := range nums {
		if _, ok := myMap[num]; ok {
			myMap[num] = append(myMap[num], idx)
		} else {
			myMap[num] = []int{idx}
		}
	}
	for key, idxList := range myMap {
		temp := target - key
		if _, ok := myMap[temp]; ok {
			result[0] = idxList[0]
			if temp == key && len(idxList) > 1 {
				result[1] = idxList[1]
				break
			} else if temp != key {
				result[1] = myMap[temp][0]
				break
			}
		}
	}
	sort.Ints(result)
	return result
}
